TFNetworkLayer.py

```diff
@@ -1043,53 +1047,237 @@ def batch_norm(self, data,
+        momentum=momentum, epsilon=epsilon,
+        sample_mean=sample_mean, sample_variance=sample_variance,
+        gamma=gamma, beta=beta,
+        masked_time=masked_time)
+
     with tf.variable_scope("batch_norm"):
+      input_data = data.copy()
       if masked_time:
-        x = data.get_placeholder_flattened(keep_dims=True)
+        # Default case
+        axes = data.get_axes_from_description(axes)
+        assert len(axes) == 1 and axes[0] == data.feature_dim_axis
+        x = input_data.get_placeholder_flattened(keep_dims=True)
         mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+        variables_shape = input_data.get_bc_spatial_batch_shape()
+        mean = tf.reshape(mean, shape=variables_shape)
+        variance = tf.reshape(variance, shape=variables_shape)
       else:
-        x = data.placeholder
-        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
+        # Perform batch_norm operations in NHWC format in order to
+        # keep consistency when saving/loading the variables.
+        if data.is_batch_feature_major:
+          # Transform the input to NHWC
+          input_data = data.copy_as_batch_spatial_major()
+        axes = input_data.get_axes_from_description(axes)
+        assert input_data.feature_dim_axis in axes
+        assert input_data.batch_dim_axis not in axes  # Not supported
+        assert input_data.time_dim_axis not in axes  # Not supported
+        x = input_data.placeholder
+        axes_norm = axes
+        axes_stat = list(set(input_data.get_axes()) - set(axes_norm))
+        opts = dict(zip(axes_norm, [-1] * len(axes_norm)))
+        variables_shape = input_data.get_bc_shape(opts)
+        mean, variance = tf.nn.moments(x, axes=axes_stat, keep_dims=True)
       if sample_mean is None:
         with self.var_creation_scope():
           sample_mean = self.add_param(tf.Variable(
-            initial_value=tf.zeros(data.get_bc_spatial_batch_shape()),
+            initial_value=tf.zeros(variables_shape),
             name="%s_%s_mean" % (self.name, data.name),
-            trainable=False))
+            trainable=True))
         # Use exponential moving average of batch mean.
         # Note: We could also use cumulative moving average. Our Theano implementation does that for inference.
         sample_mean = tf.assign_add(sample_mean, (mean - sample_mean) * momentum)
+      else:
+        sample_mean = tf.reshape(sample_mean, shape=variables_shape)
       if sample_variance is None:
         # Note: Our Theano implementation does not use a moving average for this.
         with self.var_creation_scope():
           sample_variance = self.add_param(tf.Variable(
-            initial_value=tf.ones(data.get_bc_spatial_batch_shape()),
+            initial_value=tf.ones(variables_shape),
             name="%s_%s_variance" % (self.name, data.name),
-            trainable=False))
+            trainable=True))
```

Comment:
Why the change to `trainable=True`? This is wrong, or not? It is not a variable which you train. The update happens because you have `tf.assign_add` below.

Reply:
If it's `False`,  these variables are not "seen" by the updater even if the custom update is set.

Reply:
Fixed